services/dedup/badger/badger.go

```diff
@@ -101,20 +101,23 @@ func (d *BadgerDB) Get(key string) (int64, bool, error) {
 
 func (d *BadgerDB) Set(kvs []types.KeyValue) error {
 	defer d.stats.NewTaggedStat("dedup_commit_duration_seconds", stats.TimerType, stats.Tags{"mode": "badger"}).RecordDuration()()
+	return d.set(kvs)
+}
 
+func (d *BadgerDB) set(kvs []types.KeyValue) error {
 	txn := d.badgerDB.NewTransaction(true)
-	for _, message := range kvs {
+	defer txn.Discard()
+	for i := range kvs {
+		message := kvs[i]
 		value := strconv.FormatInt(message.Value, 10)
 		e := badger.NewEntry([]byte(message.Key), []byte(value)).WithTTL(d.window.Load())
 		err := txn.SetEntry(e)
-		if err == badger.ErrTxnTooBig {
+		if errors.Is(err, badger.ErrTxnTooBig) {
 			if err = txn.Commit(); err != nil {
 				return err
 			}
-			txn = d.badgerDB.NewTransaction(true)
-			if err = txn.SetEntry(e); err != nil {
-				return err
-			}
+			txn.Discard()
+			return d.set(kvs[i:])
```

Comment:
We might end up creating a big recursive stack if everything ends up with `badger.ErrTxnTooBig` the following error. We can still re-use the existing logic while making sure that Discard is being called for the transaction. wdyt?

```GO
func (d *BadgerDB) set(kvs []types.KeyValue) error {
	txn := d.badgerDB.NewTransaction(true)
	defer txn.Discard() // Discard will clean up the last transaction, even if there's an early return.

	for i := 0; i < len(kvs); i++ {
		message := kvs[i]
		value := strconv.FormatInt(message.Value, 10)
		e := badger.NewEntry([]byte(message.Key), []byte(value)).WithTTL(d.window.Load())
		err := txn.SetEntry(e)

		if errors.Is(err, badger.ErrTxnTooBig) {
			if err := txn.Commit(); err != nil {
				return err
			}
			
			// Discard the old transaction and create a new one
			txn.Discard()
			txn = d.badgerDB.NewTransaction(true)
			if err = txn.SetEntry(e); err != nil {
				return err
			}
		} else if err != nil {
			return err
		}
	}
	return txn.Commit()
}
```

Reply:
should we go with a `writeBatch` instead?(refer `enterprise/suppress-user/internal/badgerdb/badgerdb.go`)

Reply:
How are we going to handle `badger.ErrTxnTooBig `in that case, since writeBatch would do things in a single transaction? Currently, we are creating a new transaction when getting `badger.ErrTxnTooBig`?

Reply:
writeBatch handles it internally
if it gets too big, it commits and starts a new transaction.
We can avoid all this handling on our side

Reply:
TIL. Didn't find anything on the API but this function provides the context.

```GO
func (wb *WriteBatch) handleEntry(e *Entry) error {
	if err := wb.txn.SetEntry(e); err != ErrTxnTooBig {
		return err
	}
	// Txn has reached it's zenith. Commit now.
	if cerr := wb.commit(); cerr != nil {
		return cerr
	}
	// This time the error must not be ErrTxnTooBig, otherwise, we make the
	// error permanent.
	if err := wb.txn.SetEntry(e); err != nil {
		wb.err.Store(err)
		return err
	}
	return nil
}
```